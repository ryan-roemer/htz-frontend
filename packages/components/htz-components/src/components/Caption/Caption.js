import React from 'react';
import PropTypes from 'prop-types';
import { FelaComponent, } from 'react-fela';
import {
  parseComponentProp,
  parseStyleProps,
  parseTypographyProp,
} from '@haaretz/htz-css-tools';
import setColor from '../../utils/setColor';
import { stylesPropType, } from '../../propTypes/stylesPropType';

const captionWrapperStyle = ({
  theme,
  backgroundColor,
  color,
  typeStyles,
  miscStyles,
}) => {
  const { bgc, captionColor, captionTypeSettings, fontFamily, fontWeight, } =
    theme.captionStyles || {};
  const typeSettings = typeStyles || captionTypeSettings;

  return {
    fontFamily,
    fontWeight,

    extend: [
      // Set background color
      ...(backgroundColor || bgc
        ? [
          parseComponentProp(
            'backgroundColor',
            backgroundColor || bgc,
            theme.mq,
            setColor,
            theme.color
          ),
        ]
        : []),

      // Set color
      ...(captionColor || color
        ? [
          parseComponentProp(
            'color',
            color || captionColor,
            theme.mq,
            setColor,
            theme.color
          ),
        ]
        : []),
      // set typographic styles (line height and font-size)
      ...(typeSettings ? [ parseTypographyProp(typeSettings, theme.type), ] : []),
      // Trump all other styles with those defined in `miscStyles`
      ...(miscStyles ? parseStyleProps(miscStyles, theme.mq, theme.type) : []),
    ],
  };
};

const CaptionWrapper = ({
  backgroundColor, // eslint-disable-line react/prop-types
  color, // eslint-disable-line react/prop-types
  typeStyles, // eslint-disable-line react/prop-types
  miscStyles, // eslint-disable-line react/prop-types
  children, // eslint-disable-line react/prop-types
}) => (
  <FelaComponent
    backgroundColor={backgroundColor}
    color={color}
    typeStyles={typeStyles}
    miscStyles={miscStyles}
    rule={captionWrapperStyle}
  >
    {children}
  </FelaComponent>
);

const creditRule = ({ theme, prefix, floatCredit, }) => {
  const { fontWeight, } = theme.captionStyles.creditStyles || {};
  return {
    display: 'inline',
    fontWeight,
    ':before': {
      content: `'${prefix}: '`,
    },
    extend: [ ...(floatCredit ? [ { float: 'inline-end', }, ] : []), ],
  };
};

// eslint-disable-next-line react/prop-types
const Credit = ({ prefix, floatCredit, children, }) => (
  <FelaComponent
    prefix={prefix}
    floatCredit={floatCredit}
    rule={creditRule}
    render="span"
  >
    {children}
  </FelaComponent>
);

/**
 * A Caption for Images, Embeds, Videos, etc.
 * @param {Object} props
 */
const Caption = props => {
  if (!props.caption && !props.credit) {
    return null;
  }
  return (
    <CaptionWrapper {...props}>
      <FelaComponent style={{ marginEnd: '1rem', }} render="span">
        {props.caption}
      </FelaComponent>
      {props.credit ? (
        <Credit prefix={props.creditprefix} floatCredit={props.floatCredit}>
          {props.credit}
        </Credit>
      ) : (
        ''
      )}
    </CaptionWrapper>
  );
};

Caption.propTypes = {
  /**
   * Input for the caption.
   */
  caption: PropTypes.string,
  /**
   * Input for the credit.
   */
  credit: PropTypes.string,
  /**
   * The credits prefix.
   */
  creditprefix: PropTypes.string,
  /**
   * Should the credit float.
   */
  floatCredit: PropTypes.bool,
  /**
   * The color of the background.
   */
  backgroundColor: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.array,
    PropTypes.object,
  ]),
  /**
   * The color of the inner text.
   */
  color: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.array,
    PropTypes.object,
  ]),
  /**
   * The typography of the text.
   */
  typeStyles: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.array,
    PropTypes.object,
  ]),
  /**
   * A special property holding miscellaneous CSS values that
   * trumps all default values. Processed by
   * [`parseStyleProps`](https://Haaretz.github.io/htz-frontend/htz-css-tools#parsestyleprops)
   */
  miscStyles: stylesPropType,
};

Caption.defaultProps = {
  caption: null,
  credit: null,
  creditprefix: 'קרדיט',
  floatCredit: false,
  backgroundColor: null,
  color: null,
  typeStyles: null,
  miscStyles: null,
};

export default Caption;
