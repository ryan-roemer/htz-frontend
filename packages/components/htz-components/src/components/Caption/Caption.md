<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**

- [Examples](#examples)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

### Examples

```jsx
<Grid rowSpacing={{ amount: 2, nUp: 3, }}>
  
  <GridItem width={1/3}>
    <p>With the default props</p> 
    <Caption
      caption='סתם תיאור של משהו'
      credit='מישהו'
    />
  </GridItem>
  
  <GridItem width={1/3}>
    <p>Left to right, and dynamic background</p> 
    <Caption
      caption='Caption content'
      credit='Some guy'
      creditprefix='Credit'
      direction='ltr'
      backgroundColor={[{until: 'xl', value: ['tertiary', '-3'],}, { from: 'xl', value: 'primary',}, ]}
    />
  </GridItem>
  
  <GridItem width={1/3}>
    <p>With the credit having its own direction</p> 
    <Caption
      caption='סתם תיאור של משהו'
      credit='Some guy'
      creditprefix='Credit'
      backgroundColor={['neutral', '-10']}
      color={'neutral'}
    />
  </GridItem>
</Grid>
```
