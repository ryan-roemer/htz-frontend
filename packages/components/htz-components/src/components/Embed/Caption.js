import React from 'react';
import PropTypes from 'prop-types';
import { createComponent, } from 'react-fela';

const captionWrapper = () => ({
  backgroundColor: '#2c2c2c',
  color: '#fff',
  padding: '0.375rem',
  fontWeight: '700',
  fontSize: '1.75rem',
  lineHeight: '1.5em',
  display: 'block',
  width: 'auto',
  boxSizing: 'border-box',
});

const CaptionWrapper = createComponent(captionWrapper, 'figcaption');

const credit = () => ({
  color: '#b2b2b2',
  float: 'left',
  fontWeight: '400',
  marginRight: '0.375rem',
  ':before': {
    content: '"קרדיט: "',
  },
});

const Credit = createComponent(credit, 'span', props => Object.keys(props));

const Caption = props => (
  <CaptionWrapper
    inputTemplate={props.inputTemplate}
    embedType={props.embedType}
  >
    {props.caption}
    <Credit>{props.credit}</Credit>
  </CaptionWrapper>
);

Caption.propTypes = {
  inputTemplate: PropTypes.string.isRequired,
  caption: PropTypes.string,
  credit: PropTypes.string,
  embedType: PropTypes.string.isRequired,
};

Caption.defaultProps = {
  caption: '',
  credit: '',
};

export default Caption;
