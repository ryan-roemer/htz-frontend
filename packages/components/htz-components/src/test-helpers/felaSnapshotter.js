import React from 'react';
import createRenderer from '../components/StyleProvider/createRenderer';
import StyleProvider from '../components/StyleProvider/StyleProvider';
import createFelaSnapshotter from './createFelaSnapshotter';

const testRenderer = createRenderer({ isRtl: true, });
const felaSnapshotter = createFelaSnapshotter(StyleProvider, testRenderer);

export default felaSnapshotter;
