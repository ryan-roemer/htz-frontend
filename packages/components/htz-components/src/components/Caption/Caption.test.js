import React from 'react';
import felaSnapshotter from '../../test-helpers/felaSnapshotter';
import Caption from './Caption';

it('should correctly render a Caption bar with both credit and caption ', () => {
  const snapshot = felaSnapshotter(
    <Caption caption="The Caption" credit="me" />
  );
  expect(snapshot).toMatchSnapshot();
});
it('should correctly render a Caption bar with only the caption, Left to right, and with dynamic background ', () => {
  const snapshot = felaSnapshotter(
    <Caption
      caption="The Caption"
      direction="ltr"
      bgc={[
        { until: 'xl', value: [ 'tertiary', '-3', ], },
        { from: 'xl', value: 'primary', },
      ]}
      color={[
        { until: 'xl', value: 'primary', },
        { from: 'xl', value: [ 'tertiary', '-3', ], },
      ]}
    />
  );
  expect(snapshot).toMatchSnapshot();
});
it('should correctly render a Caption bar with only the credit ', () => {
  const snapshot = felaSnapshotter(
    <Caption credit="me" creditPrefix="Credit" />
  );
  expect(snapshot).toMatchSnapshot();
});

it('should render no Caption bar', () => {
  const snapshot = felaSnapshotter(<Caption />);
  expect(snapshot).toMatchSnapshot();
});
