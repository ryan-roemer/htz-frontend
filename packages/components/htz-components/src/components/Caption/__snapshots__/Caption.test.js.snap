// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render a Caption bar with both credit and caption  1`] = `
Object {
  "component": <div>
    <div
      className="htz-a htz-b htz-c htz-d htz-e htz-f htz-g htz-h htz-i"
    >
      <span
        className="htz-j"
      >
        The Caption
      </span>
      <span
        className="htz-k htz-l htz-m htz-f htz-g htz-h htz-i"
      >
        קרדיט: me
      </span>
    </div>
  </div>,
  "styles": ".htz-a{word-break:break-word}.htz-b{overflow-x:hidden}.htz-c{font-weight:700}.htz-d{padding-right:1rem}.htz-e{color:#505050}.htz-j{margin-left:1rem}.htz-k{display:inline}.htz-l{flex-shrink:0}.htz-m{font-weight:300}@media (max-width: 79.9375em){.htz-f{font-size:2rem}.htz-g{line-height:1.5em}}@media (min-width: 80em){.htz-h{font-size:2rem}.htz-i{line-height:1.5em}}",
}
`;

exports[`should correctly render a Caption bar with only the caption, Left to right, and with dynamic background  1`] = `
Object {
  "component": <div>
    <div
      className="htz-a htz-b htz-c htz-d htz-e htz-f htz-g htz-h htz-i htz-j"
    >
      <span
        className="htz-k"
      >
        The Caption
      </span>
      
    </div>
  </div>,
  "styles": ".htz-a{word-break:break-word}.htz-b{overflow-x:hidden}.htz-c{font-weight:700}.htz-d{padding-right:1rem}.htz-k{margin-left:1rem}@media (max-width: 79.9375em){.htz-e{color:#0B7EB5}.htz-f{font-size:2rem}.htz-g{line-height:1.5em}}@media (min-width: 80em){.htz-h{color:#FFA6A6}.htz-i{font-size:2rem}.htz-j{line-height:1.5em}}",
}
`;

exports[`should correctly render a Caption bar with only the credit  1`] = `
Object {
  "component": <div>
    <div
      className="htz-a htz-b htz-c htz-d htz-e htz-f htz-g htz-h htz-i"
    >
      <span
        className="htz-j"
      />
      <span
        className="htz-k htz-l htz-m htz-f htz-g htz-h htz-i"
      >
        קרדיט: me
      </span>
    </div>
  </div>,
  "styles": ".htz-a{word-break:break-word}.htz-b{overflow-x:hidden}.htz-c{font-weight:700}.htz-d{padding-right:1rem}.htz-e{color:#505050}.htz-j{margin-left:1rem}.htz-k{display:inline}.htz-l{flex-shrink:0}.htz-m{font-weight:300}@media (max-width: 79.9375em){.htz-f{font-size:2rem}.htz-g{line-height:1.5em}}@media (min-width: 80em){.htz-h{font-size:2rem}.htz-i{line-height:1.5em}}",
}
`;

exports[`should render no Caption bar 1`] = `
Object {
  "component": <div />,
  "styles": "",
}
`;
